### ISSUE TITLE
Why don't host and basePath support path templating?

### ISSUE THREAD

**Issue by maxlinc** [OAI/OpenAPI-Specification#169]
The Swagger 2 schema says that neither host nor basePath support path templating. We probalby discussed this but I don't remember the exact reason and it seems like a big limitation. It's common to have parameters in either the host or in the basePath. You can at least work around the basePath limitation with excessively long paths, but there is no workaround for parameter in a host. Swagger will force you to hardcode a value.

Real-world examples:
- Host parameter:
  - [Rackspace](http://api.rackspace.com/): `https://{data_center}.servers.api.rackspacecloud.com/v2/{account_id}`
  - [Okta](http://developer.okta.com/docs/getting_started/design_principles.html): `https://{your_company}.okta.com/api/v1/apps/{app_id}`
  - [Zendesk](https://developer.zendesk.com/rest_api/docs/core/organizations): `https://{subdomain}.zendesk.com/api/v2/organizations.json`
- basePath parameter
  - [Apigee](http://apigee.com/docs/management/apis/get/organizations/%7Borg_name%7D): `https://api.enterprise.apigee.com/v1/organizations/{org_name}/apis`
  - [Heroku](https://devcenter.heroku.com/articles/platform-api-reference): `https://api.heroku.com/apps/{app_id}/addons`

I think the reason may have been just:
- We didn't have a plan for global parameters (though we do define _available_ global parameters which is likely good enough, because using a parameter in a host or basePath template could imply or require a $ref to the parameter name)
- We just didn't have enough time to put together a sample and tests before Swagger 2 was released

It's also worth noting that neither host nor basePath are required, nor are they required to resolve to the actual API (e.g. you could set your host to `example.com`).

I'm hoping to at least get support targeted for the next spec, if not an update to this spec clarifying that host/basePath templating is not required but may be supported by some projects.

**Comment by Jasontang501** [comment]
Echo on what maxlinc said above, provide support for parameter in a host will be great as this is a big blocker for us to adapt Swagger in our current projects.

**Comment by mohsen1** [comment]
What happens if an operation has a path parameter which is the exact same name of this proposed global path parameter in basePath?

How the tooling can decide which value goes to which parameter?

Here is an example:

```yaml
---
swagger: '2.0'
info:
  version: 1.0.0
  title: Pets Store

basePath: /{name}/
paths:
  /{name}:
    get:
      parameters:
        - name: name
          in: path
          description: name
          type: string
        
      responses:
        200:
          description: Returns all the pets

```

**Comment by maxlinc** [comment]
@mohsen1 You raise an interesting question about a case where the spec is probably ambiguous, but not something that worries me about basePath in particular. The basePath allows you to write DRY swagger, but it can always be written in a more redundant, expanded form without a basePath. And those cases are equally ambiguous.

Your example should be the equivalent of:

```yaml
---
swagger: '2.0'
info:
  version: 1.0.0
  title: Pets Store

paths:
  /{name}/{name}:
    get:
      parameters:
        - name: name
          in: path
          description: name
          type: string

      responses:
        200:
          description: Returns all the pets
```

Here's a more realistic example:

```yaml
paths:
  /users/{name}/friends/{name}
```

That's ambiguous. You shouldn't use the same variable name if you're talking about two different variables. This doesn't just confuse Swagger, it would confuse most client and server-side APIs that do parameter expansion or extraction.

For example, it would be equally bad as a rails route:

```rb
get '/users/:name/friends/:name', to: 'friends#get', as: 'friend'
```

And similarly, I don't think scoping makes it less ambiguous:
```rb
scope 'users/:name' do
  get 'friends/:name', to: 'friends#get', as: 'friend'
end
```

So basically: same name => same variable => same value substitution. If you don't want that than you should use distinct names:

```yaml
paths:
  /users/{user_name}/friends/{friend_name}
```

**Comment by zeymon** [comment]
Voting for this.

@mohsen1 & @maxlinc, while it is maybe not good practice to reference the same parameter multiple times in basePath and a path item, it should not matter though, because unlike a server route binding, code using a Swagger definition works the other way round: It _substitutes_ a parameter reference with a value rather than extracts it.

It could only be a problem if you had multiple definitions for the same parameter, but this is resolved by using the most specific definition already.

**Comment by zenexius** [comment]
Plus one for this feature.
I came across a situation where region and language are part of the base path
Something like

https://myserver.com/{region}/{lang}/products

Now if I don't use the base path then for each end point I have to use 

/{region}/{lang}/products
/{region}/{lang}/users
...
and so on

You got the idea

**Comment by samitpatel** [comment]
+1 

It would be great to have templating support for host so that we don't have to have multiple swagger definitions for development, test and production environment which uses different host.

Thanks.

**Comment by mm-gmbd** [comment]
+1

Really a big need (for me). I'm importing my swagger into Postman, and it doesn't interpret the path parameter (that is a part of the `basePath` in my swagger) as a variable. My current workaround is to change the swagger to `basePath: {{varName}}` so that Postman interprets it correctly, but I'd like to just wrap it with one set of curly braces for consistency.

Also, I understand that there will have to be work done on the Postman side for detecting that as a parameter path and ensuring that it is twice-wrapped with curly braces, but having it be officially supported in the Swagger spec would be the first step.

**Comment by cabbonizio** [comment]
+1

**Comment by whitlockjc** [comment]
This is a tooling issue in my opinion.  If you need to do preprocessing of your Swagger document, why does Swagger care how you do it?  This is my first impression on the matter.  If my opinion changes after reading the information above, I'll comment again.

**Comment by rglowzenski** [comment]
+1

**Comment by jcunningham77** [comment]
+1

**Comment by mkrufky** [comment]
+1

**Comment by gobengo** [comment]
+1 we could really use hostname parameterization!

**Comment by cabbonizio** [comment]
It would nice nice to somehow incorporate the use of "environments" to the spec (dev, int, qa, prod, etc.) and a hostname for each one.  Prior to selecting the "Try it out!" button, you can select an environment to invoke the API.   The use of Swagger right now just limits to a single host and would be nice to have one place to go to discover API documentation and test against any environment.

**Comment by webron** [comment]
@cabbonizio - well, there is a solution to that. As long as the spec is hosted in each environment, you can drop the host altogether, and the 'hosting' host will be assumed. I think leaking into environment management inside the spec itself is somewhat out of scope.

**Comment by cabbonizio** [comment]
Fair enough @webron.  Is the concept of hostname parameterization still on the table at least?

**Comment by webron** [comment]
Anything that's labeled as `OpenAPI.Next Proposal` is on the table.

**Comment by cabbonizio** [comment]
good to know...thx for the info @webron.  Looks like there's a decent number of votes.

**Comment by lieldulev** [comment]
Another vote here :)
+1

**Comment by ChiefORZ** [comment]
:+1:

**Comment by pradeep122** [comment]
:+1:

**Comment by cabbonizio** [comment]
In the meantime, I used a little markdown magic to provide clients of my services a grid letting them know each environment and host.   If you drop this into the info.description field in the Swagger 2.0 spec it renders very nicely in Swagger UI and Editor just pushes your operations down a bit.  It doesn't help with the test harness but it works for documentation purposes.  Make sure you put a pipe character prior to the below.

      ### Host Mapping By Environment  

      | Environment | Network Zone | Host                      |
      |-------------|--------------|---------------------------|
      | Production  | Internal     | sample.host.com
      | Production  | External     | sample.host.com
      | QA          | Internal     | sample.host.com
      | QA          | External     | sample.host.com
      | INT         | Internal     | sample.host.com
      | INT         | External     | sample.host.com

**Comment by webron** [comment]
Parent: #560.

**Comment by fehguy** [comment]
parent issue #574

**Comment by kmceachern** [comment]
👍   We'd like to provide access to Swagger UI both via a reverse proxy and by directly connecting to the host.

**Comment by talolierwork** [comment]
+1

**Comment by xeranic** [comment]
+1

**Comment by fesaab** [comment]
+1

**Comment by lucbelliveau** [comment]
+1

**Comment by n-sviridenko** [comment]
+1

**Comment by lcrabb** [comment]
+1
allowing for variable substitution from the request or response for the host name.  similar to #742

**Comment by webron** [comment]
The introduction of the Server Object (instead of `scheme`, `host` and `basePath`) with Server Variables covers this now.

**Comment by olebedev** [comment]
@webron, could you please give the link to it ^ ?

**Comment by darrelmiller** [comment]
@olebedev https://github.com/OAI/OpenAPI-Specification/blob/OpenAPI.next/versions/3.0.md#serverObject

**Comment by Download** [comment]
Can this be used for authorization url as well?
See https://github.com/swagger-api/swagger-ui/pull/3410
